Mage.Sets/src/mage/cards/k/KaitoDancingShadow.java

```diff
@@ -133,27 +93,37 @@ public boolean apply(Game game, Ability source) {
         if (watcher == null) {
             return false;
         }
-        TargetCreatureThatDealtCombatDamage target = new TargetCreatureThatDealtCombatDamage(0, 1, watcher.getPermanents());
+        Player damagedPlayer = game.getPlayer(targetPointer.getFirst(game, source));
         Player controller = game.getPlayer(source.getControllerId());
-        if (controller == null) {
+        if (controller == null || damagedPlayer == null) {
             return false;
         }
-        if (controller.chooseUse(outcome, "Return a creature card that dealt damage to hand?", source, game) && target.chooseTarget(Outcome.ReturnToHand, source.getControllerId(), source, game)) {
-            Card card = game.getCard(target.getFirstTarget());
+        FilterPermanent filter = new FilterPermanent();
+        filter.add(new PermanentReferenceInCollectionPredicate(
+                watcher.getPermanents(controller.getId(),damagedPlayer.getId())));
+        TargetPermanent target = new TargetPermanent(0, 1, filter, true);
```

Comment:
technically I think it should be an optional effect with minimum 1 target, though this implementation is practically the same

Reply:
True, but then you have to make the "return Y/N" decision without seeing the creatures highlighted, which seems likely to cause mistakes.

Reply:
I see your point, leave it as is then